src/nodes/chunk_dispatch/chunk_dispatch.c

```diff
@@ -148,8 +147,17 @@ ts_chunk_dispatch_get_chunk_insert_state(ChunkDispatch *dispatch, Point *point,
 		ts_set_compression_status(cis, chunk);
 
 		ts_subspace_store_add(dispatch->cache, chunk->cube, cis, destroy_chunk_insert_state);
+	}
+	else if (cis->rel->rd_id == dispatch->prev_cis_oid && cis == dispatch->prev_cis)
+	{
+		/* got the same item from cache as before */
+		cis_changed = false;
+	}
 
-		if (found && ts_chunk_is_compressed(chunk) && !ts_chunk_is_distributed(chunk))
+	if (found)
+	{
+		Chunk *chunk = ts_chunk_get_by_id(cis->chunk_id, true);
```

Comment:
This lookup is expensive to do per-row. Would be good to somehow merge with the above lookups, or cache it in the chunk insert state, or maybe check `cis->chunk_compressed` and `cis->chunk_data_nodes` first.

Reply:
For every new chunk the `cis` code above reads the chunk twice...the api contract of the `decompress_batches_for_insert` is to give a `Chunk` to it.

I think instead of storing random attributes of the `Chunk` struct - it would be better to store a more lightweight version of it - even more because down below where only pretty small part of the `Chunk` struct is actually being used - meanwhile where scankey is being constructed the relid for the hypertable was not available easily

I've changed to use `cis->...` for now and load the `Chunk` right before calling `decompress_batches_for_insert`